flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/ProcessFunctionWithCleanupState.scala

```diff
@@ -58,6 +58,10 @@ abstract class ProcessFunctionWithCleanupState[IN,OUT](queryConfig: StreamQueryC
         val cleanupTime = currentTime + maxRetentionTime
         // register timer and remember clean-up time
         ctx.timerService().registerProcessingTimeTimer(cleanupTime)
+        // delete expired timer
+        if (curCleanupTime != null) {
+          ctx.timerService().deleteProcessingTimeTimer(curCleanupTime)
```

Comment:
If we delete the timer, do we need the `needToCleanupState` any more?

Reply:
This is a good question. The `needToCleanupState` is useless in `GroupAggProcessFunction`, `RowTimeBoundedRangeOver`, etc. But it is used in `ProcTimeBoundedRangeOver`. 
In `ProcTimeBoundedRangeOver`, both state cleanup and computation will register proctime timer, so that we still need this `needToCleanupState` to judge whether to clean up the state.